doc/sonic-application-extention/sonic-application-extention-hld.md

```diff
@@ -0,0 +1,1441 @@
+is set.
+
+2. Docker container upgrades through Kubernetes.
+
+During that processes a new Docker container gets deployed and run on the switch. Kubernetes managed features introduces a state machine for a running
+container that is reflected in the STATE DB. When container starts it sets its state as "pending" and waits till there is a "ready" flag set.
+
+A state-db-watcherd is listening to those changes. If there is a need to regenerate service file and scripts it initiates sonic-package-manager
+upgrade re-generation processes. Since the image is already downloaded and running but pending, the package manifest can be read and based on manifest
+it can generate the required files. Then, the state-db-watcherd can proceed to systemctl stop old container, systemctl start new container
+where the container_action script set the container state to "ready" and the container resumes to run the applications.
+
+3. Docker container deploy through Kubernetes.
+
+The case describe the scenario when the user is setting a label via "config kube label add <key>=<value>". If labels match the manifest on the master
+the docker image gets deployed. Using the same approach, between "pending" and "ready" states an auto-generation process can happen and register the
+Docker as a new package. With that, a deployed image will become an installed SONiC Package, that can be managed in "local" mode as well.
+
+
+### 3rd party Docker images
+
+<!-- onit in toc -->
+### Installing 3rd party image as is.
+
+It is possible to install 3rd party Docker images if the installer will use a default manifest. This default manifest will have only mandatory fields
+and an auto-generated service will require only "docker" service to be started. The name of the service is derived from the name of the Package in *packages.json*.
+
+Default manifest:
+```json
+{
+  "service": {
+    "name": "<package-name>",
+    "requires": "docker",
+    "after": "docker"
+  }
+}
+```
+
+The default container run configuration have to be skipped for 3rd party Docker images. E.g. "--net=host", "-e NAMESPACE=$DEV" are not required
+for 3rd party unless. 'asic-service' field is ignored for 3rd party Docker image in this case.
+This settings have to be present in *container* properties in manifest.
+
+3rd party Docker image, as it has no knowledge of SONiC, will not meet the requirements described in section [SONiC Package]((#sonic-package)).
+Thus, for such Docker images the limitations are:
+- can be locally managed only, as the requirement for kube managed features is not met.
+
+Another possibility is to allow users to provide a manifest file URL.
+
+An example of the flow for a collectd Docker image:
+```
+admin@sonic:~$ sudo sonic-package-manager repository add collectd puckel/docker-collectd
+admin@sonic:~$ sudo sonic-package-manager install collectd --manifest https://some-server/manifests/collectd/manifest.json
+```
```

Comment:
Does the user need to update package.json for collected (3rd party docker) or it would be taken care implicitly by the sonic package manger?

Reply:
If the user wants to restrict the 3rd party docker images or non-SONiC built-in docker images to use only some % of the device storage, any option available?

Reply:
> Does the user need to update package.json for collected (3rd party docker) or it would be taken care implicitly by the sonic package manager?

User normally should not touch packages.json, this is internal to sonic-package-manager, so it is taken care implicitly

Reply:
> If the user wants to restrict the 3rd party docker images or non-SONiC built-in docker images to use only some % of the device storage, any option available?

This is a functionality that is out of scope in this document. The reason is that it is not really a feature that can be applied for non SONiC dockers or Application Extension Dockers. Ideally, we might need to plan for general support for resource constraints for all dockers that can run in the system, something like:

```
config feature storage-quota swss 50%
```